import::here(celldex,
    'BlueprintEncodeData', 'HumanPrimaryCellAtlasData',
    'DatabaseImmuneCellExpressionData', 'ImmGenData',
    'MonacoImmuneData', 'MouseRNAseqData', 'NovershternHematopoieticData'
)
import::here(Seurat,
    'ScaleData', 'RunPCA', 'RunUMAP', 'FindNeighbors', 'FindClusters'
)
import::here(DESeq2,
    'DESeqDataSetFromMatrix', 'estimateSizeFactors', 'estimateDispersions',
    'nbinomWaldTest', 'results'
)


## Objects
## celldex_switch

## Functions
## run_standard_analysis_workflow
## counts_from_seurat
## FindDESeq2Markers


#' Celldex Switch
#' 
#' @description Choose a dataset: ENCODE, HPCA, DICE, Immgen, Monaco, MouseRNAseq, Hemato
#' 
celldex_switch=list2env(list(

    #259 RNA-seq samples of pure stroma and immune cells as generated and supplied by Blueprint and ENCODE
    'ENCODE'=BlueprintEncodeData,

    #713 microarray samples from the Human Primary Cell Atlas (HPCA) (Mabbott et al., 2013).
    'HPCA'=HumanPrimaryCellAtlasData,

    #1561 bulk RNA-seq samples of sorted immune cell populations
    'DICE'=DatabaseImmuneCellExpressionData,

    # 830 microarray samples of pure mouse immune cells, generated by the Immunologic Genome Project (ImmGen)
    'ImmGen'=ImmGenData,

    #114 bulk RNA-seq samples of sorted immune cell populations that can be found in GSE107011.
    'Monaco'=MonacoImmuneData,

    #358 bulk RNA-seq samples of sorted cell populations
    'MouseRNAseq'=MouseRNAseqData,

    #211 bulk human microarray samples of sorted hematopoietic cell populations that can be found in GSE24759
    'Hemato'=NovershternHematopoieticData
))


#' Standard Analysis Workflow
#'
#' @references
#' \href{https://satijalab.org/seurat/articles/integration_introduction.html#perform-an-integrated-analysis}{Seurat Documentation}
#' 
run_standard_analysis_workflow <- function(seurat_obj, ndim=30) {
    seurat_obj <- ScaleData(seurat_obj, verbose = FALSE)
    seurat_obj <- RunPCA(seurat_obj, npcs = ndim, verbose = FALSE)  # required for RunUMAP
    seurat_obj <- RunUMAP(seurat_obj, reduction = "pca", dims = 1:ndim)
    seurat_obj <- FindNeighbors(seurat_obj, reduction = "pca", dims = 1:ndim)
    seurat_obj <- FindClusters(seurat_obj, resolution = 0.5)
    return(seurat_obj)
}


#' Counts From Seurat
#' 
#' @description
#' Retrieve the counts data from a Seurat object
#' Returns genes as columns and cell barcodes as rows
#' 
#' @param seurat_obj Seurat V4 object
#' @param genes list of genes, unused genes will be filtered
#' @param nrows set a limit if just eda
#' 
#' 
counts_from_seurat <- function(
    seurat_obj,
    genes=c(),
    barcodes=c(),
    nrows=NA,
    normalized=FALSE  # if this is true, then this returns the same data as SeuratObject::FetchData
) {
    # genes
    if (length(genes)==0) {
        genes <- 1:nrow(seurat_obj)
    }

    # barcodes
    if (length(barcodes)==0) {
        if (is.na(nrows)) {
            num_cells <- ncol(seurat_obj)
            barcodes <- 1:num_cells
        } else {
            barcodes <- 1:nrows
        }
    }
    if (normalized==TRUE) {
        df <- data.frame(t(seurat_obj[genes, barcodes][['RNA']]@data))
    } else{
        df <- data.frame(t(seurat_obj[genes, barcodes][['RNA']]@counts))
    }
    
    return(df)
}


#' FindDESeq2Markers
#'
#' @description
#' This is a copy of [Seurat::DESeq2DETest()], because using [Seurat::FindMarkers()] with `test.use="DESeq2"` gives the wrong Fold Change, making the volcano appear uncentered.
#' 
#' @references
#' \href{https://github.com/satijalab/seurat/blob/656fc8b562d53e5d0cedda9e09d9dda81e8c00e9/R/differential_expression.R#L1426-L1458}{DESeq2DETest}
#' 
FindDESeq2Markers <- function(
    object,
    ident.1,
    ident.2,
    alpha=0.05,
    pAdjustMethod="bonferroni",
    verbose=TRUE
) {
    main <- function() {
        data.use <- slot(object[['RNA']], 'counts')
        data.use <- round(data.use)

        cells.1 <- WhichCells(object = object, idents = ident.1)
        cells.2 <- WhichCells(object = object, idents = ident.2)

        group.info <- data.frame(row.names = c(cells.1, cells.2))
        group.info[cells.1, "group"] <- "Group1"
        group.info[cells.2, "group"] <- "Group2"
        group.info[, "group"] <- factor(x = group.info[, "group"])
        group.info$wellKey <- rownames(x = group.info)

        dds1 <- DESeqDataSetFromMatrix(
            countData = data.use[, group.info$wellKey],
            colData = group.info,
            design = ~ group
        )
        dds1 <- estimateSizeFactors(object = dds1)
        dds1 <- estimateDispersions(object = dds1, fitType = "local")
        dds1 <- nbinomWaldTest(object = dds1)
        res <- DESeq2::results(
            object = dds1,
            contrast = c("group", "Group1", "Group2"),
            alpha = alpha,
            pAdjustMethod = pAdjustMethod
        )
    }

    if (verbose) {
        res <- main()
    } else {
        res <- suppressMessages(main())
    }
    
    return(as.data.frame(res))
}
