## Code used to analyze scRNAseq data
## Cluster each sample individually
## Based on: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
##
## This code is adapted from DIY_scRNAseq.R, which is freely available here: 
## https://diytranscriptomics.com/scripts
## See: https://github.com/harrisonized/diy-transcriptomics


wd = dirname(this.path::here())  # wd = '~/github/R/seaFish'
reticulate::use_condaenv('r-reticulate')  # required for cellassign to access tensorflow through python
library('rjson')
library('readr')
suppressMessages(library('Seurat'))
library('Matrix')
suppressMessages(library('DropletUtils'))
suppressMessages(library('dplyr'))
library('cellassign')  # takes a few seconds
library('SingleR') # automated cell type annotation ('label transfer') using reference data
suppressMessages(library('celldex')) # a large collection of reference expression datasets with curated cell type labels for use with SingleR package
library('scran')
library('pheatmap')
library('ggplot2')
library('optparse')
library('logr')
source(file.path(wd, 'R', 'functions', 'file_io.R'))  # read_10x
source(file.path(wd, 'R', 'functions', 'list_tools.R'))  # multiple_replacement


# ----------------------------------------------------------------------
# Pre-script settings

# args
option_list = list(
    make_option(c("-i", "--input-dir"),
                default='data/ballesteros-2020/input',
                metavar='data/ballesteros-2020/input',
                type="character",
                help="directory of directories (one up) containing standard 10x files: barcodes.tsv, genes.tsv, and matrix.mtx"),

    make_option(c("-c", "--celldex"),
                default='ImmGen', metavar='ImmGen', type="character",
                help="Choose from: ['ENCODE', 'HPCA', 'DICE', 'ImmGen', 'Monaco', 'MouseRNAseq', 'Hemato']"),

    make_option(c("-e", "--ensembl"),  default=FALSE,
                metavar="FALSE", action="store_true", type="logical",
                help="When querying celldex"),

    make_option(c("-g", "--gene-of-interest"), default="Dnase1l1",
                metavar="Dnase1l1", type="character",
                help="choose a gene"),

    make_option(c("-t", "--troubleshooting"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="enable if troubleshooting to prevent overwriting your files")
)
opt_parser <- OptionParser(option_list=option_list)
opt <- parse_args(opt_parser)
troubleshooting <- opt[['troubleshooting']]
figures_dir <- multiple_replacement(
    opt[['input-dir']], c('data'='figures', 'input'='output')
)

# Start Log
start_time <- Sys.time()
log <- log_open(paste0("cluster_samples-",
                       strftime(start_time, format="%Y%m%d_%H%M%S"), '.log'))
log_print(paste('Script started at:', start_time))


switch=list2env(list(

    #259 RNA-seq samples of pure stroma and immune cells as generated and supplied by Blueprint and ENCODE
    'ENCODE'=BlueprintEncodeData,

    #713 microarray samples from the Human Primary Cell Atlas (HPCA) (Mabbott et al., 2013).
    'HPCA'=HumanPrimaryCellAtlasData,

    #1561 bulk RNA-seq samples of sorted immune cell populations
    'DICE'=DatabaseImmuneCellExpressionData,

    # 830 microarray samples of pure mouse immune cells, generated by the Immunologic Genome Project (ImmGen)
    'ImmGen'=ImmGenData,

    #114 bulk RNA-seq samples of sorted immune cell populations that can be found in GSE107011.
    'Monaco'=MonacoImmuneData,

    #358 bulk RNA-seq samples of sorted cell populations
    'MouseRNAseq'=MouseRNAseqData,

    #211 bulk human microarray samples of sorted hematopoietic cell populations that can be found in GSE24759
    'Hemato'=NovershternHematopoieticData
))


# ----------------------------------------------------------------------
# Main

samples <- list.dirs(file.path(wd, opt[['input-dir']]), full.names=FALSE, recursive=FALSE)
log_print(paste(Sys.time(),
                'Found', length(samples), 'samples...',
                paste(samples, collapse=', '))
)

for (sample_name in samples) {

    tryCatch({

        # ----------------------------------------------------------------------
        # Read Data

        log_print(paste(Sys.time(), 'Reading', sample_name, '...'))

        expr_mtx <- read_10x(file.path(wd, opt[['input-dir']], sample_name))
        seurat_obj <- CreateSeuratObject(counts = expr_mtx, min.cells = 3) %>% 
            NormalizeData(verbose = FALSE) %>% 
            FindVariableFeatures(verbose = FALSE)

        # QC plot
        VlnPlot(seurat_obj, c("nCount_RNA", "nFeature_RNA"), pt.size = 0.1)
        if (!troubleshooting) {
            ggsave(file.path(wd, figures_dir, sample_name, 'qc',
                             paste0('violin-', sample_name, '-raw_qc.png')),
                   height=800, width=1200, dpi=300, units="px", scaling=0.5)
        }

        # ----------------------------------------------------------------------
        # Filter data

        # Note: emptyDrops cannot be run twice
        # For the ballesteros-2020 dataset, this has already been run
        # drop_stats <- DropletUtils::emptyDrops(expr_mtx)

        seurat_obj <- subset(
            seurat_obj,
            subset = ((nCount_RNA < 20000) & 
                      (nCount_RNA > 1000) & 
                      (nFeature_RNA > 1000))
        )

        # another QC plot
        ggplot(seurat_obj@meta.data, aes(nCount_RNA, nFeature_RNA)) +
            geom_point(alpha = 0.7, size = 0.5) +
            labs(x = "Total UMI counts per cell", y = "Number of genes detected")
        if (!troubleshooting) {
            ggsave(file.path(wd, figures_dir, sample_name, 'qc',
                             paste0('scatter-', sample_name, '-num_genes_vs_counts.png')),
                   height=800, width=1200, dpi=300, units="px", scaling=0.5)
        }


        # ----------------------------------------------------------------------
        # Standard Clustering Workflow

        log_print(paste(Sys.time(), 'Running standard clustering workflow...'))

        seurat_obj <- ScaleData(seurat_obj, verbose = FALSE)
        seurat_obj <- RunPCA(seurat_obj, npcs = 40, verbose = FALSE)
        seurat_obj <- RunUMAP(seurat_obj, reduction = "pca", dims = 1:40)
        seurat_obj <- FindNeighbors(seurat_obj, reduction = "pca", dims = 1:40)
        seurat_obj <- FindClusters(seurat_obj, resolution = 0.5)

        # Plot UMAP
        fig <- DimPlot(
            seurat_obj, reduction = "umap", split.by = "orig.ident",
            label = TRUE
        ) +
        theme(
            strip.text.x = element_blank(),
            plot.title = element_text(hjust = 0.5)
        ) +
        ggtitle("Unlabeled Clusters")

        if (!troubleshooting) {
            ggsave(file.path(wd, figures_dir, sample_name,
                             paste0('umap-', sample_name, '-clusters.png')),
                   plot = fig,
                   height=800, width=800, dpi=300, units="px", scaling=0.5)
        }

        # Plot UMAP for specific gene of interest
        gene = opt[['gene-of-interest']]  # gene = 'IGHM'
        fig <- FeaturePlot(
            seurat_obj, reduction = "umap", features = c(gene),
            pt.size = 0.4,
            min.cutoff = 'q10',
            # split.by = "orig.ident",
            order = TRUE, label = FALSE
        )
        if (!troubleshooting) {
            ggsave(file.path(wd, figures_dir, sample_name,
                             paste0('umap-', sample_name, '-', tolower(gene), '.png')),
                   plot = fig,
                   height=800, width=800, dpi=300, units="px", scaling=0.5)
        }

        # ----------------------------------------------------------------------
        # Label clusters

        log_print(paste(Sys.time(), 'Labeling clusters...'))

        # note: this overwrites the labels generated by the standard clustering workflow
        sce_counts <- as.SingleCellExperiment(seurat_obj)
        get_data=switch[[opt$celldex]]
        ref_data <- get_data(ensembl = opt[['ensembl']])  # default: ensembl=FALSE
        predictions <- SingleR(
            test=sce_counts,
            assay.type.test=1,
            ref=ref_data,
            labels=ref_data[['label.main']]
        )
        sce_counts[["cell_type"]] <- predictions[['labels']]

        # not needed yet
        # plotScoreHeatmap(predictions)

        # Plot UMAP
        seurat_counts <- as.Seurat(sce_counts, counts = NULL)
        fig <- DimPlot(
            seurat_counts, reduction = "UMAP", 
            group.by = "cell_type",
            # split.by = "treatment",  # facet if necessary
            label = FALSE
        ) + ggtitle("Labeled Clusters")

        if (!troubleshooting) {
            ggsave(file.path(
                       wd, figures_dir, sample_name,
                       paste0('umap-', sample_name, '-', tolower(opt[['celldex']]), '_labeled.png')
                   ),
                   plot = fig,
                   height=800, width=1000, dpi=300, units="px", scaling=0.5)
        }
    },

    # pass
    error = function(condition) {
        log_print(paste("WARNING: SAMPLE", sample_name, "NOT PROCESSED!!!"))
        log_print(paste("Error message: ", conditionMessage(condition)))
        NULL
    })

}

end_time <- Sys.time()
log_print(paste('Script ended at:', Sys.time()))
log_print(paste("Script completed in:", difftime(end_time, start_time)))
log_close()
